src/main/java/io/cdap/plugin/gcp/datastore/source/DatastoreRecordReader.java

```diff
@@ -62,21 +45,33 @@
 
   private static final Logger LOG = LoggerFactory.getLogger(DatastoreRecordReader.class);
 
-  private QueryResults<Entity> results;
+  private Iterator<EntityResult> results;
   private Entity entity;
   private long index;
   private LongWritable key;
 
   @Override
-  public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptContext) {
+  public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptContext) throws IOException {
     Configuration config = taskAttemptContext.getConfiguration();
-    Query<Entity> query = transformPbQuery(((QueryInputSplit) inputSplit).getQuery(), config);
-    Datastore datastore = DatastoreUtil.getDatastore(
+    Query query = ((QueryInputSplit) inputSplit).getQuery();
+    Datastore datastore = DatastoreUtil.getDatastoreV1(
       config.get(DatastoreSourceConstants.CONFIG_SERVICE_ACCOUNT),
       config.getBoolean(DatastoreSourceConstants.CONFIG_SERVICE_ACCOUNT_IS_FILE, true),
       config.get(DatastoreSourceConstants.CONFIG_PROJECT));
     LOG.trace("Executing query split: {}", query);
-    results = datastore.run(query);
+    RunQueryRequest request = RunQueryRequest.newBuilder()
+      .setQuery(query)
+      .setPartitionId(PartitionId.newBuilder()
+                        .setNamespaceId(config.get(DatastoreSourceConstants.CONFIG_NAMESPACE))
+                        .setProjectId(config.get(DatastoreSourceConstants.CONFIG_PROJECT)))
+      .build();
+    try {
+      QueryResultBatch batch = datastore.runQuery(request).getBatch();
+      taskAttemptContext.getCounter(FileInputFormatCounter.BYTES_READ).increment(batch.getSerializedSize());
+      results = batch.getEntityResultsList().iterator();
+    } catch (DatastoreException e) {
```

Comment:
Old code throws [com.google.cloud.datastore.DatastoreException](https://googleapis.dev/java/google-cloud-datastore/latest/com/google/cloud/datastore/Datastore.html#run-com.google.cloud.datastore.Query-com.google.cloud.datastore.ReadOption...-).

I throw an IOException here because RecordReader.initialize() throws [IOException, InterruptedException](https://hadoop.apache.org/docs/current/api/org/apache/hadoop/mapreduce/RecordReader.html#initialize-org.apache.hadoop.mapreduce.InputSplit-org.apache.hadoop.mapreduce.TaskAttemptContext-).